### PR TITLE
Feature/structure query

### PR DESCRIPTION
**Pull request by TasDeniz**
save structured query for abide

### CODE REVIEW

File: src/main/java/de/samply/share/broker/rest/InquiryHandler.java [hunk 1]
```diff
@@ -127,9 +123,9 @@ static boolean isQueryLanguageCql(String queryLanguage) {
    */
   public int storeAndRelease(String simpleQueryDtoJson, int userid, String inquiryName,
       String inquiryDescription, int exposeId, int voteId, List<String> resultTypes,
-      boolean bypassExamination, boolean isMonitoring) {
+      boolean bypassExamination, boolean isMonitoring, boolean structuredQuery) {
```
Comment (alexanderkiel) [inline src/main/java/de/samply/share/broker/rest/InquiryHandler.java:126]: Please don't add another boolean toggle. Use a query type instead and remove also isMonitoring.

File: src/main/java/de/samply/share/broker/rest/InquiryHandler.java [hunk 2]
```diff
@@ -250,6 +230,32 @@ private int store(String query, int userid, String inquiryName,
     return returnValue;
   }
 
+  private void storeCriteria(String query, boolean isMonitoring, boolean structuredQuery,
+      Inquiry inquiry, Connection connection) {
+    if (structuredQuery) {
```
Comment (alexanderkiel) [inline src/main/java/de/samply/share/broker/rest/InquiryHandler.java:235]: Just switch on the query type here. Or even better: put the query into a class that has a method for saving the criteria = real Java polymorphism.

File: src/main/java/de/samply/share/broker/rest/Monitoring.java [hunk 3]
```diff
@@ -221,18 +222,18 @@ public Response createQuery(@HeaderParam(HttpHeaders.AUTHORIZATION) String auth,
       @Parameter(
           name = "query",
           description = "Query as a JSON object",
-          schema = @Schema(implementation = QueryObject.class))
-          QueryObject queryObject) {
+          schema = @Schema(implementation = QueryContainer.class))
+          QueryContainer<CqlQuery[]> queryContainer) {
```
Comment (alexanderkiel) [inline src/main/java/de/samply/share/broker/rest/Monitoring.java:226]: I'm sure we sad, that this endpoint should be kept until the monitoring solution is changed. So please introduce a new endpoint for the QueryContainer and remove this endpoint after the monitoring is changed.

File: src/main/java/de/samply/share/broker/rest/Monitoring.java [hunk 4]
```diff
@@ -248,13 +249,13 @@ public Response createQuery(@HeaderParam(HttpHeaders.AUTHORIZATION) String auth,
   public Response getResult(@HeaderParam(HttpHeaders.AUTHORIZATION) String auth,
       @PathParam("id") int queryId) {
     try {
-      if (auth == null || !checkBasicAuth(auth)) {
+      if (auth == null || !checkBasicAuth(auth,"icinga.username", "icinga.password")) {
```
Comment (alexanderkiel) [inline src/main/java/de/samply/share/broker/rest/Monitoring.java:252]: This is unrelated to the structured query feature. Please put that into it's own PR.

File: src/main/java/de/samply/share/broker/rest/Searchbroker.java [hunk 5]
```diff
@@ -1271,8 +1274,63 @@ public Response setSiteByName(@HeaderParam(HttpHeaders.AUTHORIZATION) String aut
     }
   }
 
+  /**
+   * Save structured query.
```
Comment (alexanderkiel) [inline src/main/java/de/samply/share/broker/rest/Searchbroker.java:1278]: Please bring the documentation inline with the endpoint and method name.

File: src/main/java/de/samply/share/broker/rest/Searchbroker.java [hunk 6]
```diff
@@ -1271,8 +1274,63 @@ public Response setSiteByName(@HeaderParam(HttpHeaders.AUTHORIZATION) String aut
...
+  @APIResponses({
+      @APIResponse(
+          responseCode = "200",
+          description = "ok",
+          content = @Content(
+              mediaType = MediaType.APPLICATION_JSON,
+              schema = @Schema(implementation = String.class))),
+      @APIResponse(responseCode = "500", description = "Internal Server Error")
+  })
+  @Operation(summary = "Save query in searchbroker database")
```
Comment (alexanderkiel) [inline src/main/java/de/samply/share/broker/rest/Searchbroker.java:1295]: Please bring the documentation inline with the endpoint and method name.

File: src/main/java/de/samply/share/broker/rest/Searchbroker.java [hunk 7]
```diff
@@ -1271,8 +1274,63 @@ public Response setSiteByName(@HeaderParam(HttpHeaders.AUTHORIZATION) String aut
...
+  public Response executeSQ(@HeaderParam(HttpHeaders.AUTHORIZATION) String auth,
+      @Parameter(
+          name = "structuredQuery",
```
Comment (alexanderkiel) [inline src/main/java/de/samply/share/broker/rest/Searchbroker.java:1298]: This isn't a structured query. It's a query container that contains the structured query and the targets.

File: src/main/java/de/samply/share/broker/rest/Searchbroker.java [hunk 8]
```diff
@@ -1271,8 +1274,63 @@ public Response setSiteByName(@HeaderParam(HttpHeaders.AUTHORIZATION) String aut
...
+          description = "Query as a JSON object",
```
Comment (alexanderkiel) [inline src/main/java/de/samply/share/broker/rest/Searchbroker.java:1299]: Not true, see above.

File: src/main/java/de/samply/share/broker/rest/Searchbroker.java [hunk 9]
```diff
@@ -1271,8 +1274,63 @@ public Response setSiteByName(@HeaderParam(HttpHeaders.AUTHORIZATION) String aut
...
+          schema = @Schema(implementation = QueryContainer.class))
+          QueryContainer<String> queryContainer) {
+    try {
+      if (auth == null || !checkBasicAuth(auth, "abide.username", "abide.password")) {
```
Comment (alexanderkiel) [inline src/main/java/de/samply/share/broker/rest/Searchbroker.java:1303]: We don't like to have project names in the code. Accepting a structured query is a feature that it independent of a concrete project like abide. Why do we have a separate account for executing structured queries?

File: src/main/java/de/samply/share/broker/rest/Searchbroker.java [hunk 10]
```diff
@@ -1271,8 +1274,63 @@ public Response setSiteByName(@HeaderParam(HttpHeaders.AUTHORIZATION) String aut
...
+              queryContainer.getTarget(), queryContainer.getQueryName(), true)).build();
+    } catch (Exception e) {
+      LOGGER.warn(e.getMessage());
+      return Response.status(Response.Status.INTERNAL_SERVER_ERROR).build();
+    }
+  }
 
 
+  /**
+   * Save structured query.
```
Comment (alexanderkiel) [inline src/main/java/de/samply/share/broker/rest/Searchbroker.java:1317]: Docu not in line with code.

File: src/main/java/de/samply/share/broker/rest/Searchbroker.java [hunk 11]
```diff
@@ -1271,8 +1274,63 @@ public Response setSiteByName(@HeaderParam(HttpHeaders.AUTHORIZATION) String aut
...
+   *
+   * @return 200 or 500 code
```
Comment (alexanderkiel) [inline src/main/java/de/samply/share/broker/rest/Searchbroker.java:1319]: First a Response is returned and not a code. Second, the code is 204 and not 200. Third, it's better to explain in which cases which response is returned and why.

File: src/main/java/de/samply/share/broker/rest/Searchbroker.java [hunk 12]
```diff
@@ -1271,8 +1274,63 @@ public Response setSiteByName(@HeaderParam(HttpHeaders.AUTHORIZATION) String aut
...
+  @OPTIONS
+  @Path("/executeSQ")
+  @Produces(MediaType.TEXT_PLAIN)
+  @Consumes({MediaType.APPLICATION_JSON, MediaType.TEXT_PLAIN})
+  @APIResponses({
+      @APIResponse(responseCode = "204", description = "no-content")
+  })
+  @Operation(summary = "Save structured query in searchbroker database (OPTIONS for CORS)")
+  public Response executeSqOptions() {
+    LOGGER.info("createSQ called (OPTIONS)");
```
Comment (alexanderkiel) [inline src/main/java/de/samply/share/broker/rest/Searchbroker.java:1330]: Do we usually log every request with info level?

File: src/main/resources/db/migration/V009__structuredQuery.sql [hunk 13]
```diff
@@ -0,0 +1 @@
+ALTER TYPE INQUIRY_CRITERIA_TYPE ADD VALUE 'IC_SQ';
```
Comment (alexanderkiel) [inline src/main/resources/db/migration/V009__structuredQuery.sql:1]: Can we use IC_STRUCTURED_QUERY instead? I don't know if we know that SQ is in 6 month. Also This needs a newline at the end of the file.

File: src/main/webapp/WEB-INF/conf/samply.share.broker.conf [hunk 14]
```diff
@@ -8,8 +8,12 @@ mdr.url=https://mdr.ccp-it.dktk.dkfz.de/v3/api/mdr/
...
+icinga.path=/v1/actions/process-check-report
 icinga.username=dktk-searchbroker-java
 icinga.password=EQrfn3WW226ChnXbV4JYhgIPhQSc
 # (A) for dev system (B) for production
-icinga.site_suffix=GBA (A)
+icinga.site_suffix=BBMRI (A)
+icinga.project=-bbmri
+
+abide.username=test123
+abide.password=test123
```
Comment (alexanderkiel) [inline src/main/webapp/WEB-INF/conf/samply.share.broker.conf:19]: Do we really need to have a fixed password here? Please explain the reason for the separate account.

**Review by alexanderkiel — CHANGES_REQUESTED** [review]

**Comment by alexanderkiel** [comment]
And please, rebase this on develop. Currently it's on master.

File: src/main/webapp/WEB-INF/conf/samply.share.broker.conf (hunk 14 above)
Comment (TasDeniz) [reply src/main/webapp/WEB-INF/conf/samply.share.broker.conf:19]: Because Icinga is an diffrent api and software. And it is better to use different credentials in different projects?

File: src/main/java/de/samply/share/broker/rest/Searchbroker.java (hunk 12 above)
Comment (TasDeniz) [reply src/main/java/de/samply/share/broker/rest/Searchbroker.java:1330]: then debug?

File: src/main/java/de/samply/share/broker/rest/Monitoring.java (hunk 4 above)
Comment (TasDeniz) [reply src/main/java/de/samply/share/broker/rest/Monitoring.java:252]: So it is better to dublicate the code only to remove it later again?

Comment (alexanderkiel) [reply src/main/java/de/samply/share/broker/rest/Monitoring.java:292]: Doing so is the only way to be compatible with the current monitoring solution. Otherwise an update of the search broker would break the monitoring. The code is also not duplicated, because it's different.

File: src/main/webapp/WEB-INF/conf/samply.share.broker.conf (hunk 14 above)
Comment (alexanderkiel) [reply src/main/webapp/WEB-INF/conf/samply.share.broker.conf:19]: We could call the credentials structured-query-execution.username and .password. But why are the credentials not managed like the other credentials. Why do we have them fixed in the config files and not editable in the UI? Or are only credentials in the UI that are used by the share-client in order to access other API's?

File: changelog.md [hunk 15]
```diff
@@ -4,9 +4,16 @@ All notable changes to this project will be documented in this file.
...
+### Added
+- Save "Structured-Query"
+### Changed
+- For the monitoring blackbox test it is necessary to add a query name in the json object
+
+
 ## [8.4.2 - 2021-12-16]
 ### Security
-- Update Log4J to v2.16.0
+- Update Log4J to v2.16.0 ([#29](https://github.com/samply/searchbroker/pull/29))
```
Comment (alexanderkiel) [inline changelog.md:16]: 👍

File: src/docker/samply.share.broker.conf [hunk 16]
```diff
@@ -14,3 +14,6 @@ icinga.password={icinga-password}
 # (A) for dev system (B) for production
 icinga.site_suffix={icinga-site-suffix}
 icinga.project={icinga-project}
+
+abide.username={abide-username}
```
Comment (alexanderkiel) [inline src/docker/samply.share.broker.conf:18]: structured-query-execution.username

File: src/main/java/de/samply/share/broker/filter/BasicAuth.java [hunk 17]
```diff
@@ -0,0 +1,27 @@
+package de.samply.share.broker.filter;
+
+import de.samply.share.common.utils.ProjectInfo;
+import java.nio.charset.StandardCharsets;
+import java.util.Base64;
+
+public class BasicAuth {
```
Comment (alexanderkiel) [inline src/main/java/de/samply/share/broker/filter/BasicAuth.java:7]: Can't we use some standard JAX-RS authentication filter here?

File: src/main/java/de/samply/share/broker/model/QueryContainer.java [hunk 18]
```diff
@@ -0,0 +1,35 @@
+package de.samply.share.broker.model;
+
+import java.util.List;
+
+public class QueryContainer<T> {
```
Comment (alexanderkiel) [inline src/main/java/de/samply/share/broker/model/QueryContainer.java:5]: Please add at least a class JavaDoc. Isn't hat enforced through Google style?

File: src/main/java/de/samply/share/broker/filter/BasicAuth.java (hunk 17 above)
Comment (alexanderkiel) [reply src/main/java/de/samply/share/broker/filter/BasicAuth.java:7]: And if not, please write at least a test.

File: src/main/java/de/samply/share/broker/rest/InquiryHandler.java [hunk 19]
```diff
@@ -250,6 +229,32 @@ private int store(String query, int userid, String inquiryName,
     return returnValue;
   }
 
+  private void storeCriteria(String query, InquiryCriteriaType queryType,
+      Inquiry inquiry, Connection connection) {
+    if (queryType.getLiteral().equals(InquiryCriteriaType.IC_STRUCTURED_QUERY.getLiteral())) {
```
Comment (alexanderkiel) [inline src/main/java/de/samply/share/broker/rest/InquiryHandler.java:234]: You can just use switch on the enum directly.

File: src/main/java/de/samply/share/broker/rest/Monitoring.java [hunk 20]
```diff
@@ -224,18 +226,57 @@ public Response createQuery(@HeaderParam(HttpHeaders.AUTHORIZATION) String auth,
           schema = @Schema(implementation = QueryObject.class))
           QueryObject queryObject) {
     try {
-      if (auth == null || !checkBasicAuth(auth)) {
+      if (auth == null || !checkBasicAuth(auth, "icinga.username", "icinga.password")) {
```
Comment (alexanderkiel) [inline src/main/java/de/samply/share/broker/rest/Monitoring.java:229]: Also here, I would call that monitoring.username and monitoring.password because the API we provide is for monitoring and Icinga is only one specific user of our common monitoring endpoint.

File: src/main/java/de/samply/share/broker/rest/Monitoring.java [hunk 21]
```diff
@@ -224,18 +226,57 @@ public Response createQuery(@HeaderParam(HttpHeaders.AUTHORIZATION) String auth,
...
         return Response.status(Response.Status.UNAUTHORIZED).build();
       }
       return Response.ok(
           SearchController.releaseQuery(new Gson().toJson(queryObject.getCqlQueryList()),
-              queryObject.getTarget())).build();
+              queryObject.getTarget(), "Icinga Test", InquiryCriteriaType.IC_CQL)).build();
```
Comment (alexanderkiel) [inline src/main/java/de/samply/share/broker/rest/Monitoring.java:234]: I would also call this Monitoring and not Icinga Test.

File: src/main/java/de/samply/share/broker/rest/Monitoring.java [hunk 22]
```diff
@@ -224,18 +226,57 @@ public Response createQuery(@HeaderParam(HttpHeaders.AUTHORIZATION) String auth,
...
 
+  /**
+   * Create a new monitoring query.
+   * @param queryContainer the query as cql string and the target sites as list
+   * @return response if the query has been created
+   */
+  @Path("/v2/queries")
+  @POST
+  @Produces(MediaType.TEXT_PLAIN)
+  @Consumes({MediaType.APPLICATION_JSON, MediaType.TEXT_PLAIN})
```
Comment (alexanderkiel) [inline src/main/java/de/samply/share/broker/rest/Monitoring.java:249]: It can't consume text/plain.

File: src/main/java/de/samply/share/broker/rest/Monitoring.java [hunk 23]
```diff
@@ -224,18 +226,57 @@ public Response createQuery(@HeaderParam(HttpHeaders.AUTHORIZATION) String auth,
...
+  @APIResponses({
+      @APIResponse(
+          responseCode = "200",
+          description = "ok",
+          content = @Content(
+              mediaType = MediaType.APPLICATION_JSON,
```
Comment (alexanderkiel) [inline src/main/java/de/samply/share/broker/rest/Monitoring.java:255]: Here you say, that re 200 response is in JSON and in @Produces you say, its text/plain.

File: src/main/java/de/samply/share/broker/rest/Searchbroker.java [hunk 24]
```diff
@@ -1271,7 +1275,63 @@ public Response setSiteByName(@HeaderParam(HttpHeaders.AUTHORIZATION) String aut
...
   }
 
+  /**
+   * Execute structured query.
+   *
+   * @param queryContainer the structuredQuery and the targets
+   * @return response if the query has been created
+   */
+  @POST
+  @Path("/executeSQ")
```
Comment (alexanderkiel) [inline src/main/java/de/samply/share/broker/rest/Searchbroker.java:1285]: I think, I changed my mind. The monitoring endpoint uses `/queries` and `/v2/queries` with create semantics of CRUD. So we should do the same for the structured-query. Maybe `/structured-query/queries`.

File: src/main/java/de/samply/share/broker/rest/Searchbroker.java [hunk 25]
```diff
@@ -1271,7 +1275,63 @@ public Response setSiteByName(@HeaderParam(HttpHeaders.AUTHORIZATION) String aut
...
+    try {
+      if (auth == null || !checkBasicAuth(auth, "abide.username", "abide.password")) {
+        return Response.status(Response.Status.UNAUTHORIZED).build();
+      }
+      return Response.ok(
+          SearchController.releaseQuery(queryContainer.getQuery(),
+              queryContainer.getTarget(), queryContainer.getQueryName(),
+              InquiryCriteriaType.IC_STRUCTURED_QUERY)).build();
+    } catch (Exception e) {
+      LOGGER.warn(e.getMessage());
```
Comment (alexanderkiel) [inline src/main/java/de/samply/share/broker/rest/Searchbroker.java:1313]: This is an error and not a warning.

File: src/main/java/de/samply/share/broker/rest/Searchbroker.java [hunk 26]
```diff
@@ -1271,7 +1275,63 @@ public Response setSiteByName(@HeaderParam(HttpHeaders.AUTHORIZATION) String aut
...
+      return Response.status(Response.Status.INTERNAL_SERVER_ERROR).build();
```
Comment (alexanderkiel) [inline src/main/java/de/samply/share/broker/rest/Searchbroker.java:1314]: Also append the exception message to the response.

File: src/main/resources/db/migration/V009__structuredQuery.sql [hunk 27]
```diff
@@ -0,0 +1,2 @@
+ALTER TYPE INQUIRY_CRITERIA_TYPE ADD VALUE 'IC_STRUCTURED_QUERY';
+ALTER TYPE INQUIRY_CRITERIA_TYPE ADD VALUE 'IC_DEFAULT';
```
Comment (alexanderkiel) [inline src/main/resources/db/migration/V009__structuredQuery.sql:2]: Please add a newline at the end of file.

**Review by alexanderkiel — CHANGES_REQUESTED** [review]

File: src/main/java/de/samply/share/broker/rest/Monitoring.java [hunk 28]
```diff
@@ -226,11 +227,9 @@ public Response createQuery(@HeaderParam(HttpHeaders.AUTHORIZATION) String auth,
           schema = @Schema(implementation = QueryObject.class))
           QueryObject queryObject) {
     try {
-      if (auth == null || !checkBasicAuth(auth, "icinga.username", "icinga.password")) {
-        return Response.status(Response.Status.UNAUTHORIZED).build();
-      }
       return Response.ok(
-          SearchController.releaseQuery(new Gson().toJson(queryObject.getCqlQueryList()),
+          SearchController.releaseQuery(new Gson().toJson(queryObject.getCqlQueryList().getQueries()
+                  .toArray(new CqlQuery[0])),
```
Comment (alexanderkiel) [inline src/main/java/de/samply/share/broker/rest/Monitoring.java:232]: Why do we need an array here?

File: src/main/java/de/samply/share/broker/rest/Monitoring.java [hunk 29]
```diff
@@ -243,6 +242,7 @@ public Response createQuery(@HeaderParam(HttpHeaders.AUTHORIZATION) String auth,
    * @param queryContainer the query as cql string and the target sites as list
    * @return response if the query has been created
    */
+  @Secured
```
Comment (alexanderkiel) [inline src/main/java/de/samply/share/broker/rest/Monitoring.java:245]: Please use a separate annotation for the monitoring account (separate filter) here.

File: src/main/java/de/samply/share/broker/filter/AuthenticationFilter.java [hunk 30]
```diff
@@ -33,7 +35,8 @@
 public class AuthenticationFilter implements ContainerRequestFilter {
```
Comment (alexanderkiel) [inline src/main/java/de/samply/share/broker/filter/AuthenticationFilter.java:35]: Please split this filter into one for the JWT based authentication as before and one for the monitoring. This filter is just way to complex. And please add tests for both filters.

File: src/main/java/de/samply/share/broker/rest/Searchbroker.java [hunk 31]
```diff
@@ -1297,20 +1296,17 @@ public Response setSiteByName(@HeaderParam(HttpHeaders.AUTHORIZATION) String aut
   @Operation(summary = "Save query in searchbroker database")
   public Response executeSQ(@HeaderParam(HttpHeaders.AUTHORIZATION) String auth,
```
Comment (alexanderkiel) [inline src/main/java/de/samply/share/broker/rest/Searchbroker.java:1297]: It's now saveStructuredQuery again. Sorry. :-)

File: src/main/java/de/samply/share/broker/rest/Searchbroker.java [hunk 32]
```diff
@@ -1321,15 +1317,15 @@ public Response executeSQ(@HeaderParam(HttpHeaders.AUTHORIZATION) String auth,
...
   @OPTIONS
-  @Path("/executeSQ")
+  @Path("/structured-query/queries")
   @Produces(MediaType.TEXT_PLAIN)
   @Consumes({MediaType.APPLICATION_JSON, MediaType.TEXT_PLAIN})
   @APIResponses({
       @APIResponse(responseCode = "200", description = "Ok, when query is released")
   })
   @Operation(summary = "Save structured query in searchbroker database (OPTIONS for CORS)")
   public Response executeSqOptions() {
```
Comment (alexanderkiel) [inline src/main/java/de/samply/share/broker/rest/Searchbroker.java:1327]: saveStructuredQueryOptions

File: src/main/java/de/samply/share/broker/rest/Searchbroker.java [hunk 33]
```diff
@@ -1281,8 +1279,9 @@ public Response setSiteByName(@HeaderParam(HttpHeaders.AUTHORIZATION) String aut
    * @param queryContainer the structuredQuery and the targets
    * @return response if the query has been created
    */
+  @Secured
```
Comment (alexanderkiel) [inline src/main/java/de/samply/share/broker/rest/Searchbroker.java:1282]: What kind of authentication we like to use here? The monitoring (icing) authentication is not appropriate. Why not the Basic Auth with a separate structured query account?

**Review by alexanderkiel — CHANGES_REQUESTED** [review]

File: src/main/java/de/samply/share/broker/rest/Monitoring.java (hunk 28 above)
Comment (TasDeniz) [reply src/main/java/de/samply/share/broker/rest/Monitoring.java:232]: because i changed in the storeCriteria method in the class InquiryHandler.
see src/main/java/de/samply/share/broker/rest/InquiryHandler.java

File: src/main/java/de/samply/share/broker/filter/BasicAuthPermission.java [hunk 34]
```diff
@@ -0,0 +1,9 @@
+package de.samply.share.broker.filter;
+
+/**
+ * Created on 21.12.2018.
```
Comment (alexanderkiel) [inline src/main/java/de/samply/share/broker/filter/BasicAuthPermission.java:4]: Comment is misleading :-)

File: src/main/java/de/samply/share/broker/filter/BasicAuthenticationFilter.java [hunk 35]
```diff
@@ -0,0 +1,84 @@
...
+      }
+    }
+  }
+
+  private boolean checkPermissions(List<BasicAuthPermission> allowedAccessPermissions,
+      String authHeader) throws UnsupportedEncodingException {
+    String[] tokens = (new String(Base64.getDecoder().decode(authHeader.split(" ")[1]),
+        "UTF-8")).split(":");
+    final String username = tokens[0];
+    final String password = tokens[1];
```
Comment (alexanderkiel) [inline src/main/java/de/samply/share/broker/filter/BasicAuthenticationFilter.java:65]: This will fail if there are no two parts after the split.

File: src/main/java/de/samply/share/broker/filter/BasicAuthSecure.java [hunk 36]
```diff
@@ -0,0 +1,24 @@
...
+ */
+@NameBinding
+@Retention(RUNTIME)
+@Target({TYPE, METHOD})
+public @interface BasicAuthSecure {
+  /**
+   * Get the access permissions roles.
+   * @return the access permissions roles
+   */
+  BasicAuthPermission[] value() default {};
```
Comment (alexanderkiel) [inline src/main/java/de/samply/share/broker/filter/BasicAuthSecure.java:23]: I would put only one premising here, because you check only the last anyway.

File: src/main/java/de/samply/share/broker/filter/BasicAuthenticationFilter.java [hunk 37]
```diff
@@ -0,0 +1,84 @@
...
+    String usernameConfig = "";
+    String passwordConfig = "";
+    for (BasicAuthPermission basicAuthPermission : allowedAccessPermissions) {
+      if (BasicAuthPermission.STRUCTURED_QUERY.equals(
+          basicAuthPermission)) {
+        usernameConfig = ProjectInfo.INSTANCE.getConfig().getProperty("structured.query.username");
```
Comment (alexanderkiel) [inline src/main/java/de/samply/share/broker/filter/BasicAuthenticationFilter.java:71]: please put the config name in the BasicAuthPermission Enum, so that you don't need an if statement here.

File: src/main/java/de/samply/share/broker/rest/Monitoring.java (hunk 28 above)
Comment (alexanderkiel) [reply src/main/java/de/samply/share/broker/rest/Monitoring.java:233]: So the query string as argument in releaseQuery is a serialised JSON now? I wouldn't do that. I would't pass JSON as serialised string between Java objects. Why not use a generic JSON object or POJO here and serialise it right before it goes into the DB.

**Review by alexanderkiel — CHANGES_REQUESTED** [review]

**Review by alexanderkiel — APPROVED** [review]
